rx_signal_processing/rx_dsp_chain.cu

```diff
@@ -15,406 +15,238 @@
-    radarctrl_socket.bind("ipc:///tmp/feeds/2");
+    gpuErrchk(cudaStreamAddCallback(dp->get_cuda_stream(),
+                  DSPCore::initial_memcpy_callback, dp, 0));
 
-    zmq::socket_t ack_socket(sig_proc_context, ZMQ_PAIR);
-    ack_socket.bind("ipc:///tmp/feeds/3"); // REVIEW #37 check the errno on all binds
+    call_decimate<DecimationType::bandpass>(dp->get_rf_samples_p(),
+      dp->get_first_stage_output_p(),
+      dp->get_first_stage_bp_filters_p(), first_stage_dm_rate,
+      rx_metadata.numberofreceivesamples(), filters.get_first_stage_lowpass_taps().size(), rx_freqs.size(),
+      total_antennas, "First stage of decimation", dp->get_cuda_stream());
 
-    zmq::socket_t timing_socket(sig_proc_context, ZMQ_PAIR);
-    timing_socket.bind("ipc:///tmp/feeds/4");
 
-    auto gpu_properties = get_gpu_properties();
-    print_gpu_properties(gpu_properties);
+    // When decimating, we go from one set of samples for each antenna in the first stage
+    // to multiple sets of reduced samples for each frequency in further stages. Output samples are
+    // grouped by frequency with all samples for each antenna following each other
+    // before samples of another frequency start. In the first stage need a filter for each frequency,
+    // but in the next stages we only need one filter for all data sets.
+    dp->allocate_and_copy_second_stage_filter(filters.get_second_stage_lowpass_taps().data(),
+                                                filters.get_second_stage_lowpass_taps().size());
 
-    uint32_t first_stage_dm_rate = 0, second_stage_dm_rate = 0, third_stage_dm_rate = 0; // REVIEW #32 this line only initializes third stage, the others are only declared. bad style?
-                                                                                         // REVIEW just a mistake. This doesnt work like python does
-    //Check for non integer dm rates
-    if (fmod(rx_rate,sig_options.get_first_stage_sample_rate()) > 0.0) {
-        //TODO(keith): handle error
-    }
-    else if (fmod(sig_options.get_first_stage_sample_rate(),
-                    sig_options.get_second_stage_sample_rate()) > 0.0) {
-        //TODO(keith): handle error
-    }
-    else if(fmod(sig_options.get_second_stage_sample_rate(),
-                sig_options.get_third_stage_sample_rate()) > 0.0) { // REVIEW #1 or #2 perhaps a comment needed to clarify what this error means (dm rate needs to be int)
-        //TODO(keith): handle error                                     // REPLY okay
-    }
-    else{
-        auto float_dm_rate = rx_rate/sig_options.get_first_stage_sample_rate(); // REVIEW #26 don't like the name rate_f, not obvious why it's called that, seems like a temp variable, so perhaps indicate that?
-        first_stage_dm_rate = static_cast<uint32_t>(float_dm_rate);             // REPLY okay
+    auto num_output_samples_2 = rx_freqs.size() * num_output_samples_1 / second_stage_dm_rate;
```

Comment:
Do you need rx_freqs.size() here since it's already included in the num_output_samples_1 variable?